log/log.go

```diff
@@ -30,18 +32,80 @@ var levelPrefix = [...]string{
 	LevelFatal:    "[FATAL] ",
 }
 
+var levelPriority = [...]syslog.Priority{
+	LevelDebug:    syslog.LOG_DEBUG,
+	LevelInfo:     syslog.LOG_INFO,
+	LevelWarning:  syslog.LOG_WARNING,
+	LevelError:    syslog.LOG_ERR,
+	LevelCritical: syslog.LOG_CRIT,
+	LevelFatal:    syslog.LOG_EMERG,
+}
+
 // Level stores the current logging level.
-var Level = LevelDebug
+var Level = LevelInfo
+
+var (
+	// SysLoggers maps each logging level to a *log.Logger that will be used for it.
+	SysLoggers = make([]*golog.Logger, LevelFatal+1)
+	// UseSyslog controls if syslog is used instead of stdout/err.
+	UseSyslog bool
+	// SyslogNetwork controls the network protocol syslog uses.
+	SyslogNetwork string
+	// SyslogRaddr is the address of a remote syslog instance.
+	SyslogRaddr string
+	// SyslogTag is the tag to pass to syslog.
+	SyslogTag string
+)
+
+func init() {
+	flag.IntVar(&Level, "loglevel", LevelInfo, "Log level (0 = DEBUG, 5 = FATAL)")
+	flag.BoolVar(&UseSyslog, "syslog", false, "Whether or not to use syslog logging")
+	flag.StringVar(&SyslogNetwork, "syslog-network", "", "Syslog network to use. If left blank along with remote address, will default to local.")
+	flag.StringVar(&SyslogRaddr, "syslog-remote", "", "Syslog server to use. Defaults to local on empty string.")
+	flag.StringVar(&SyslogTag, "syslog-tag", "", "Syslog tag to use. Will use the name of the executable if left blank.")
+}
+
+func getSysLogger(l int) *golog.Logger {
+	if SysLoggers[l] != nil {
+		return SysLoggers[l]
+	}
+
+	net := SyslogNetwork
+	// This block keeps makes this use local syslog if nothing else is configured.
+	if SyslogRaddr != "" && net == "" {
+		net = "udp"
+	}
+	w, err := syslog.Dial(net, SyslogRaddr, levelPriority[l], SyslogTag)
+	if err != nil {
+		golog.Println("Unable to dial syslog:", err)
+		return nil
+	}
+
+	SysLoggers[l] = golog.New(w, levelPrefix[l], 0)
+	return SysLoggers[l]
+}
+
+func logHelper(l int, msg string) {
+	if UseSyslog {
+		if logger := getSysLogger(l); logger != nil {
+			logger.Print(msg)
+			return
+		}
+	}
+	golog.Print(msg)
```

Comment:
Ideally we'd like to be able to use syslogging as an alternative to stdout logging. That is, if we have syslog enabled, we'd like stdout logging to be disabled. The reason: We want to additionally collect stdout/stderr using operating system facilities, but treat it as critical level, because panics get printed there. If stdout gets a duplicate of what goes to syslog, that means we can't collect it without getting duplicates.